frontends/p4/sideEffects.h

```diff
@@ -83,6 +83,46 @@ class SideEffects : public Inspector {
         expression->apply(se);
         return se.nodeWithSideEffect != nullptr;
     }
+    /// @return true if the expression may have side-effects.
+    static bool hasSideEffect(const IR::Expression* exp,
+                      ReferenceMap* refMap,
+                      TypeMap* typeMap) {
+        auto mce = exp->to<IR::MethodCallExpression>();
+        if (mce == nullptr)
+            return false;
+        // mce does not produce a side effect in the two cases:
+        //  * isValid()
+        //  * function with all in parameters
+        if (mce->method->is<IR::Member>()) {
+            auto mem = mce->method->to<IR::Member>();
+            auto basetype = typeMap->getType(mem->expr);
+            if (basetype == nullptr) {
+                basetype = mem->expr->type;
+            }
+            if (auto sc = basetype->to<IR::Type_SpecializedCanonical>())
+                basetype = sc->baseType;
+            if (basetype->is<IR::Type_HeaderUnion>()) {
+                if (mem->member == IR::Type_Header::isValid)
+                    return false;
+            } else if (basetype->is<IR::Type_Header>()) {
+                if (mem->member == IR::Type_Header::isValid)
+                    return false;
+            }
+        } else if (mce->method->is<IR::PathExpression>()) {
+            auto pe = mce->method->to<IR::PathExpression>();
+            auto decl = refMap->getDeclaration(pe->path, true);
+            if (decl->is<IR::Function>()) {
+                auto mi = MethodInstance::resolve(mce, refMap, typeMap);
```

Comment:
This is not precise enough, you are missing several cases:
- If you invoke an action or a table.apply() it may have side-effects.
- some external functions may have side-effects that are not visible to the data-plane. For these you may want to check annotations too.
- evaluating the function's arguments may have side-effects

Reply:
Than you for quick review. You are right that actions and table.apply() may produce side-effects, but hasSideEffect function returns true in all cases except  isValid() case and case with function for all in arguments.
There is a test simplify_method_calls.p4 that I have added, there is some simple action that produce side effect.

Reply:
`if( decl->is<IR::Function>())` this condition will not be met in log_msg (and other extern methods or functions).
log_msg (and other extern methods or functions) will be met in following condition
`if(decl->is<IR::Method>())`
(but I didn't include that condition because all conditions except the first one return true).
You can look at the tests with log_msg (logging.p4 and issue2201-bmv2.p4).
Method wasn't removed from program and test wasn't changed.

Reply:
since this case is actually covered, is there anything else I should change? @mbudiu-vmw